Atomic.md

```diff
@@ -0,0 +1,278 @@
+Atomic placeholder operations in VFSForGit on MacOS
+===================================================
+
+A potential hole in our current design is in our creation of
+placeholders. Today, when creating a placeholder directory, at a high
+level, we go through the following workflow:
+
+-   Create placeholder folder/file in the virtualization root
+
+-   'Mark up' the file with the proper flags/xattrs
+
+-   (If creating a file) Update the mode of the file
+
+This could result in scenarios where a incomplete placeholder appears in
+the virtualization root. The full set of failure modes is documented
+below:
+
+WritePlaceholderDirectory:
+
+-   mkdir() fails to create the placeholder directory
+
+    -   The enumeration fails as a directory wasn't created. This should
+         be retriable as a second attempt should see the IsEmpty bit on
+         the parent folder and call WritePlaceholderDirectory again. No
+         incomplete placeholders were left in the parent directory.
+
+-   We crash between mkdir() and InitializeEmptyPlaceholder
+
+    -   Placeholder folder does not have flags for virtualization root and
+        empty.
+
+    -   This folder will never be expanded as the kext will exit when
+         we do not see the InVirtualizationRoot bit
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Folder is marked as in virtualization root but not empty.
+
+        -   This folder will never be expanded as the kext will exit
+             when we do not see the IsEmpty bit
```

Comment:
And to be clear, this will manifest as a directory that is allowed to be enumerated, but appears empty. In other words, it will look like files are missing from the repo.